nemo/collections/multimodal/models/multimodal_llm/kosmos/megatron_kosmos_model.py

```diff
@@ -0,0 +1,1129 @@
+            return [self.model]
+
+    def set_inference_config(self, inference_config):
+        self._inference_config = inference_config
+
+    def get_inference_config(self):
+        return self._inference_config
+
+    def model_provider_func(self, pre_process, post_process):
+        """Model depends on pipeline paralellism."""
+        media_start_id = self.tokenizer.token_to_id(self.cfg.media_start_token)
+        media_end_id = self.tokenizer.token_to_id(self.cfg.media_end_token)
+
+        model = KosmosModel(
+            model_cfg=self.cfg,
+            vocab_size=self.padded_vocab_size,
+            media_start_id=media_start_id,
+            media_end_id=media_end_id,
+            pre_process=pre_process,
+            post_process=post_process,
+        )
+
+        # Freeze vit
+        model.vision_encoder.freeze()
+
+        logging.info(
+            f"Kosmos model initialized with {sum(p.numel() for p in model.parameters() if p.requires_grad)} trainable parameters"
+        )
+
+        return model
+
+    def forward(self, tokens, text_position_ids, attention_mask, labels, media=None):
+        output_tensor = self.model(tokens, text_position_ids, attention_mask, labels=labels, media=media)
+        return output_tensor
+
+    def fwd_bwd_step(self, dataloader_iter, batch_idx, forward_only):
+
+        tensor_shape = [self.cfg.llm.encoder_seq_length, self.cfg.micro_batch_size, self.cfg.llm.hidden_size]
+
+        # handle asynchronous grad reduction
+        no_sync_func = None
+        grad_sync_func = None
+        param_sync_func = None
+        if not forward_only and self.with_distributed_adam:
+            no_sync_func = partial(self._optimizer.no_sync, greedy_grad_copy=self.megatron_amp_O2,)
+            grad_sync_func = self.reduce_overlap_gradients
+            param_sync_func = self.sync_overlap_parameters
+
+        # run forward and backwards passes for an entire global batch
+        # we do this inside training_step to support pipeline parallelism
+        fwd_bwd_function = get_forward_backward_func()
+
+        # TODO @akhattar: remove sync related stuff from config, add num_micro_batches_with_partial_activation_checkpoints when ready
+        losses_reduced_per_micro_batch = fwd_bwd_function(
+            forward_step_func=self.get_forward_output_and_loss_func(),
+            data_iterator=dataloader_iter,
+            model=[self.model],
+            num_microbatches=get_num_microbatches(),
+            forward_only=forward_only,
+            tensor_shape=tensor_shape,
+            dtype=self.autocast_dtype,
+            grad_scaler=self.trainer.precision_plugin.scaler.scale
+            if self.cfg.precision in [16, '16', '16-mixed']
+            else None,
+            sequence_parallel=self.cfg.get('sequence_parallel', False),
+            enable_autocast=self.enable_autocast,
+            no_sync_func=no_sync_func,
+            grad_sync_func=grad_sync_func,
+            param_sync_func=param_sync_func,
```

Comment:
I don't think this will work with latest mcore, these args were moved to ModelParallelconfig